src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Rules/AttributesMustMatch.cs

```diff
@@ -74,6 +74,13 @@ public AttributesMustMatch(IRuleSettings settings, IRuleRegistrationContext cont
             ImmutableArray<AttributeData> right,
             IList<CompatDifference> differences)
         {
+            // Filter out attributes that should be excluded based on the settings.
+            if (_settings.AttributeDataSymbolFilter is not null)
+            {
+                left = left.Where(attributeData => attributeData.AttributeClass is not null && _settings.AttributeDataSymbolFilter.Include(attributeData.AttributeClass)).ToImmutableArray();
```

Comment:
Is there a need to realize this as an ImmutableArray?  I see later on we create another ImmutableArray in ExcludeNonVisibleOutsideOfAssembly.  One way we could avoid this double-pass would be to create an aggregate filter for attributes.  So make the attribute filter a CompositeSymbolFilter of the DocIDFilter and the regular SymbolFilter.

Reply:
In other words, it's good to have a filter dedicated to filtering attributes.  Instead of making that filter *just* for exclusions, make it for both visibility and exclusions.  That saves you from doing multiple filtering passes.